### ISSUE TITLE
`testFail*` does not report success when a Tester proxy is used to trigger a throw

### ISSUE THREAD

**Issue by nmushegian** [nexusdev/dapple#175]

**Comment by nmushegian** [comment]
related  https://github.com/nexusdev/dapple/pull/168

**Comment by nmushegian** [comment]
we could use a test to confirm that fallback functions don't swallow exceptions (revive `dappsys/lang`)

**Comment by ramosi** [comment]
A simplified example showing the throw not being caught by the test if called through the proxy.

The following contract - testthrow.sol
```js
contract TestThrow {
    function TestThrow() {}
    function crash() {
        throw;
    }
}
```

Using the following test contract - testthrow_test.sol
```js
import 'dapple/test.sol';
import 'testthrow.sol';

contract TestThrowTest is Test {
    TestThrow testThrow;
    Tester tester1;

    function setUp() {
        testThrow = new TestThrow();
        tester1 = new Tester();
        tester1._target(testThrow);
    }

    function testThrowDirect() {
        testThrow.crash();
    }

    function testThrowProxy() {
        TestThrow(tester1).crash();
    }
}
```

Results in the following output when running dapple test:
```
Testing...
Using local solc installation...

TestThrowTest
  test throw direct
  Passed!

  test throw proxy
  Failed!


Failing Tests
  TestThrowTest: test throw proxy
  

Summary
  Failed 1 out of 2 tests.
```

**Comment by nmushegian** [comment]
What surprises me is that adding a typed proxy function fixes it:

https://github.com/nexusdev/dapple/pull/168/files#diff-55d1d1e876b31bf3f11d6b68bdf30783R32

**Comment by mhhf** [comment]
Its not the Tester contract, its the generic call statement: `_t.call(msg.data)` which don't seem to propagate errors.

consider this minimal example:
```
import 'dapple/test.sol';


contract TestFail is Test {
  function testFailSuccess() {
    this.crash();
  }
  function testLogFailing() {
    this.call(bytes4(bytes32(sha3("crash()"))));
  }
  function crash() {
    throw;
  }
}
```

This is maybe an error in the current EtherSim implementation, we should wait for the port of testrpc and test if this have been resolved.

**Comment by nmushegian** [comment]
Throw works by performing an invalid jump which forces the contract OOG. So while `call` does swallow the exception, it should be instantly rethrown unless you explicitly limit the gas with `.call.gas(g)(data)`.

**Comment by nmushegian** [comment]
So yes, let's get off of ethereum and see if it fixes the issue

**Comment by nmushegian** [comment]
*ethersim

**Comment by mhhf** [comment]
Actually this is not a bug, but expected behavior, since throw; is a solidity based function, and solidity also handles the throw propagation. To resolve this, we have to come up with some different Tester contract by either reconstructing the function call with assembly or a different approach on users.

**Comment by nmushegian** [comment]
I'm not convinced... That doesn't explain my comment above: Unless you are deliberately restricting gas using `.call.gas(g)()`, then it will not have gas to do anything once it "returns" from the sub-call (because invalid jump consumes all the gas).

It also doesn't explain why adding a function with types fixes the problem.

**Comment by mhhf** [comment]
> It also doesn't explain why adding a function with types fixes the problem.

it fixes it because with the typed tester the tested function is called in the solidity way and not in the generic way. Solidity in the background generate a rethrown if it observes a throw which is returned as a parameter in the call opcode.
```
contract ProxyTester is Tester {
...
    function set(bytes32 key, bytes32 value) {
        MyRegistry(_t).set(key, value);
    }
}
```
Calles via:
```
MyRegistry(proxy_tester).set("no", "stop");
```

> Unless you are deliberately restricting gas using .call.gas(g)(), then it will not have gas to do anything once it "returns" from the sub-call (because invalid jump consumes all the gas).

This is true. However, I remember reading that solidity splits gas intelligently and don't passes all available gas during a call. This can be circumvent if we force passing all gas to the generic call.
See here: http://beta.dapphub.io/sandbox/1b90604846

Also chriseth just responded on this: 
> call returns true/false depending on whether the call worked or not

So we can make something along the lines out of Tester (I've still to test it with return values):
```
contract Tester {
  address _t;
  function _target( address target ) {
    _t = target;
  }
  function() {
    if(!_t.call(msg.data)) throw;
  }
}
```

**Comment by nmushegian** [comment]
This means I misunderstood how .call works. This means there's some dappsys bugs: https://github.com/nexusdev/dappsys/blob/master/contracts/actor/base.sol#L10

Your solution is good, and we should make a convention around it.

**Comment by mhhf** [comment]
fixed in #231
